python/dgl/data/movielens.py

```diff
@@ -0,0 +1,640 @@
+    "Children",
+    "Comedy",
+    "Crime",
+    "Documentary",
+    "Drama",
+    "Fantasy",
+    "Film-Noir",
+    "Horror",
+    "Musical",
+    "Mystery",
+    "Romance",
+    "Sci-Fi",
+    "Thriller",
+    "War",
+    "Western",
+]
+GENRES_ML_1M = GENRES_ML_100K[1:]
+GENRES_ML_10M = GENRES_ML_100K + ["IMAX"]
+
+try:
+    import torch
+except ImportError:
+    HAS_TORCH = False
+else:
+    HAS_TORCH = True
+
+def check_pytorch():
+    """Check if PyTorch is the backend."""
+    if HAS_TORCH is False:
+        raise ModuleNotFoundError(
+            "MovieLensDataset requires PyTorch to be the backend."
+        )
+
+class MovieLensDataset(DGLDataset):
+    r"""MovieLens dataset for edge prediction tasks. The raw datasets are extracted from 
+    `MovieLens <https://grouplens.org/datasets/movielens/>`, introduced by
+    `Movielens unplugged: experiences with an occasionally connected recommender system <https://dl.acm.org/doi/10.1145/604045.604094>`.
+
+    The datasets consist of user ratings for movies and incorporate additional user/movie information in the form of features. 
+    The nodes represent users and movies, and the edges store ratings that users assign to movies.
+
+    Statistics:
+
+    MovieLens-100K (ml-100k)
+
+    - Users: 943
+    - Movies: 1,682
+    - Ratings: 100,000 (1, 2, 3, 4, 5)
+
+    MovieLens-1M (ml-1m)
+
+    - Users: 6,040
+    - Movies: 3,706
+    - Ratings: 1,000,209 (1, 2, 3, 4, 5)
+
+    MovieLens-10M (ml-10m)
+
+    - Users: 69,878 
+    - Movies: 10,677
+    - Ratings: 10,000,054 (0.5, 1, 1.5, ..., 4.5, 5.0)
+
+    Parameters
+    ----------
+    name: str
+        Dataset name. (:obj:`"ml-100k"`, :obj:`"ml-1m"`, :obj:`"ml-10m"`). 
+    valid_ratio: int
+        Ratio of validation samples out of the whole dataset. Should be in (0.0, 1.0).
+    test_ratio: int
+        Ratio of testing samples out of the whole dataset. Should be in (0.0, 1.0) for ml-1m and ml-10m. This parameter is invalid
+        when :obj:`name` is :obj:`"ml-100k"`, since it testing samples are pre-specified.
+        Default: None
+    raw_dir : str
+        Raw file directory to download/contain the input data directory.
+        Default: ~/.dgl/
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    verbose : bool
+        Whether to print out progress information. Default: True.
+    reverse_edge : bool
+        Whether to add reverse edges in graph. Default: True.
+    transform : callable, optional
+        A transform that takes in a :class:`~dgl.DGLGraph` object and returns
+        a transformed version. The :class:`~dgl.DGLGraph` object will be
+        transformed before every access.
+
+    Attributes
+    ----------
+    train_graph: dgl.DGLGraph
+        Training graph
+    valid_graph: dgl.DGLGraph
+        Validation graph
+    test_graph: dgl.DGLGraph
+        Testing graph
+    user_feat: torch.Tensor
+        User features
+    movie_feat: torch.Tensor
+        Movie features
```

Comment:
Why not store these in `train_graph`, `valid_graph`, `test_graph`? What are their shapes?

Reply:
user_feat and movie_feat have different feature dimensions. 
ml-100k, user_feat: 23, movie_feat: 320
ml-1m, user_feat: 23, movie_feat: 319
ml-10m, user_feat: 1, movie_feat: 321

Reply:
In that case, should we return a heterogeneous graph instead?

Reply:
Emm I'm also thinking about this. Unlike classic heterogeneous graph, movielens has no complicated canonical edge types or meta-paths that might carry rich information, and the numbers of node and edge types are too small(2 and 1) to exhibit diversity. I don't think users would regard it as a heterogeneous graph. Instead, they would think it's a homogeneous graph with edge classification as the target task since the only problem that brings us to heterogeneous graph is feature dimension, which I think is a minor problem(Two MLPs to change dimensions are enough). I didn't see the necessity to make movielens a more complicated structure. Furthermore, keeping it as homogeneous graph is helpful to directly apply GNN examples in DGL to movielens (maybe with two MLPs). Heterogeneous graph would make `HeteroGraphConv` API unavoidable(I guess, may not  be true). I think users are more familiar with the former option. Hence I think keeping movielens as simple homogeneous graphs is better

Reply:
DGL's heterogeneous graph data structure is meant to cover the bipartite graphs that you mentioned. Users can choose to convert it into a homogeneous graph if they want. You can leave the issue to them. Also, don't our current examples have the issue of different feature dimensions? How do they handle this issue?

Reply:
Yeah you are right. Give me some time to refactor some codes.

Reply:
Sure. Thanks.

Reply:
fix in the latest commit. Convert graph structure to heterogeneous graphs and remove those items that are unnecessary to be stored in local.